FIPS/fip-draft_foundational_fvm.md

```diff
@@ -0,0 +1,1010 @@
+    - Syntax validation and structural validation, [as per standard](https://webassembly.github.io/spec/core/valid/index.html).
+    - No floating point instructions.
+2. Multihash the bytecode and calculate the code CID.
+3. Check if the code CID is already registered in the `ActorRegistry`.
+4. If yes, charge gas for the above operations and return the existing CID.
+5. Insert an entry in the `ActorRegistry`, binding the bytecode CID with the
+   actor spec.
+6. Return the CID, and charge the cost of the above operations, and potentially
+   a price for rent/storage.
+
+At this point, the actor will be ready to be instantiated through the standard
+`InitActor#Exec`. Any params provided will be passed through to the actor's
+constructor (method number = 0). This includes source depoyable byte streams
+(EVM bytecode) for foreign actors.
+
+#### Actor message dispatch
+
+With the exception of simple value transfers, messages trigger some actor logic
+on the recipient actor. (In the future, value transfers may also trigger
+arbitrary logic.)
+
+Today, Filecoin chain messages carry a _method number_. The current VM matches
+on recipient actor type and method number to dispatch to the appropriate
+handler. This approach is **external method dispatch**.
+
+With the FVM, actors now expose a **single entrypoint** in the form of a
+WASM-exported function, and conduct **internal method dispatch**.
+
+The signature of an actor entrypoint is:
+
+```rust
+pub fn invoke(params_block_id: u32) -> u32 { }
+```
+
+- The single uint32 argument corresponds to the block ID of the encoded input
+  parameters struct, or `0` for none.
+- The single uint32 return value corresponds to the block ID of the return data
+  struct, or `0` for none.
+
+This design decouples actors from their runtime, and is more aligned with the
+actor model, where actors interact through inboxes. It also awards maximal
+degrees of freedom and sophistication to support techniques such as structural
+pattern matching, efficient pass-through proxying, interception, and more.
+
+Because internal dispatch logic will rapidly become boilerplate, FVM SDKs should
+offer dispatch sugar.
+
+**Backwards compatibility**
+
+The _method number_ field on messages remains unchanged to preserve backwards
+compatibility. It may be deprecated and eventually removed later. Built-in
+actors continue relying on _method number_, but now perform internal dispatch.
+
+### Syscalls
+
+This section specifies all the syscalls that the IC must make available to the
+actor.
+
+Currently, the FVM charges no gas for import linkage. In the future, the FVM may charge a static or a dynamic cost for performing the linkage of the concrete functions or namespaces imported by the module.
+
+**Interface type conventions**
+
+- `off` and `len` refer to "memory offset" and "buffer length". Together, they
+  define the bounds of a slice of memory.
+- 
+- The `Result` return type 
+
+#### Namespace: `actor`
```

Comment:
I'm not sure if we should bother specifying these for now as they're designed for a "trusted" system. We'll probably be better off if we keep _this_ FIP to higher-level abstractions, introducing specific design decisions in future FIPs.

Reply:
My intention was to provide sufficient level of detail for (alternative) FVM implementations and/or SDK developers to implement off the spec. This is generally the level of detail that is submitted in equivalent documents in other communities (e.g. EIP).

Reply:
Oh, you're referring to the `actor` namespace specifically. IMO the only syscall that doesn't belong long-term is `create_actor`. Everything else will probably remain, won't it?